policy: optimize mapstate rule labels

jrajahalme opened the pull request:
Origin rule labels are ingested and processed to the selector policy, cached selectors, and mapstate without any datapath impact. These labels are metadata that is only used when policy verdicts are inspected, e.g., when exporting hubble flows. As such it would be reasonable to optimize the processing and storage cost of these labels, possibly with somewhat higher cost when extracting them.

To this end we change the origin rule label storage from the Go "slice of slices of labels" (each label consisting of three strings) to a single string. This compacts the storage of the labels, removing the need for multiple levels of slice headers and the associated heap allocated storage. The algorithm for merging sorted label arrays in string form is implemented, operating directly in string domain.

This step alone is already beneficial, but this gets a bit better by interning the string (in the 2nd commit).

Third commit changes policy rule correlation to also function in string domain, eliminating the overhead introduced by the string representation.

This PR improves the BenchmarkRegenerateCIDRDenyPolicyRules (which generates unique labels for each rule) execution time by ~5%, reduces memory consumption by ~17%, and increases the number of heap allocations by ~3%.

jrajahalme committed:
policy: Intern rule labels

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Correlate policy in string domain

Correlate policy from string format label array list to flow protobuf in
string domain. This saves some unnecessary parsing work otherwise
required when mapstate entries no longer store the LabelArrayLists in
parsed form.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

mhofstetter reviewed (approved):
`envoy` :heavy_check_mark:

kaworu reviewed (commented):
Patch LGTM but I'm deferring sig-hubble review to @gandro whom has more context about policy correlation.

gandro commented on pkg/labels/arraylist.go:
```diff
@@ -118,6 +146,20 @@ func (lsp *LabelArrayList) Merge(include ...LabelArray) LabelArrayList {
+	for j := 0; j < len(*lsp); j++ {
+		if la.Less((*lsp)[j]) {
+			*lsp = append((*lsp)[:j+1], (*lsp)[j:]...) // make space at merged[j]
+			(*lsp)[j] = la
```
nit: Note that you could use `slices.Insert` here which as some fastpaths in cases where the target slice still has available capacity.

jrajahalme replied:
Applied, thanks!

jrajahalme replied:
Turns out this was dead code left around from development of the PR, removed.

gandro commented on pkg/labels/arraylist.go:
```diff
@@ -142,3 +184,86 @@ func (lsp *LabelArrayList) MergeSorted(include LabelArrayList) LabelArrayList {
+}
+
+// merge 'b' to 'a' assuming both are sorted
+func MergeSortedLabelArrayListStrings(a, b string) string {
```
I wonder if this could benefit from some unit tests. Or is the logic covered by existing tests indirectly?

jrajahalme replied:
Will add unit tests, as it looks like current test do not cover this.

gandro reviewed (approved):
Neat, thank you!

jrajahalme committed:
policy: Store rule labels as strings

Origin rule labels are ingested and processed to the selector policy,
cached selectors, and mapstate without any datapath impact. These labels
are metadata that is only used when policy verdicts are inspected, e.g.,
when exporting hubble flows. As such it would be reasonable to optimize
the processing and storage cost of these labels, possibly with somewhat
higher cost when extracting them.

To this end we change the origin rule label storage from the Go "slice of
slices of labels" (each label consisting of three strings) to a single
(sorted) string. This compacts the storage of the labels, removing the
need for multiple levels of slice headers and the associated heap
allocated storage. The algorithm for merging sorted label arrays in
string form is implemented, operating directly in string to string
domain.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
